src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java

```diff
@@ -333,58 +338,106 @@ public static RedisConfiguration createRedisConfiguration(RedisURI redisUri) {
-				createConnectionProvider(client, LettuceReactiveRedisConnection.CODEC));
+			this.client = createClient();
 
-		if (isClusterAware()) {
+			this.connectionProvider = new ExceptionTranslatingConnectionProvider(createConnectionProvider(client, CODEC));
+			this.reactiveConnectionProvider = new ExceptionTranslatingConnectionProvider(
+					createConnectionProvider(client, LettuceReactiveRedisConnection.CODEC));
 
-			this.clusterCommandExecutor = new ClusterCommandExecutor(
-					new LettuceClusterTopologyProvider((RedisClusterClient) client),
-					new LettuceClusterConnection.LettuceClusterNodeResourceProvider(this.connectionProvider),
-					EXCEPTION_TRANSLATION);
-		}
+			if (isClusterAware()) {
+
+				this.clusterCommandExecutor = new ClusterCommandExecutor(
+						new LettuceClusterTopologyProvider((RedisClusterClient) client),
+						new LettuceClusterConnection.LettuceClusterNodeResourceProvider(this.connectionProvider),
+						EXCEPTION_TRANSLATION);
+			}
+
+			state.set(State.STARTED);
 
-		this.initialized = true;
+			if (getEagerInitialization() && getShareNativeConnection()) {
+				initConnection();
+			}
+		}
+	}
 
-		if (getEagerInitialization() && getShareNativeConnection()) {
-			initConnection();
+	@Override
+	public void stop() {
+
+		if (state.compareAndSet(State.STARTED, State.STOPPING)) {
+			resetConnection();
+			dispose(connectionProvider);
+			dispose(reactiveConnectionProvider);
+			this.client.close();
+			state.set(State.STOPPED);
 		}
 	}
 
+	@Override
+	public boolean isRunning() {
+		return State.STARTED.equals(state.get());
+	}
+
+	@Override
+	public void afterPropertiesSet() {
+		// customization hook. initialization happens in start
+	}
+
+	@Override
 	public void destroy() {
 
-		resetConnection();
+		if (State.STOPPED.equals(state.get())) {
+			if (clusterCommandExecutor != null) {
 
-		if (clusterCommandExecutor != null) {
+				try {
+					clusterCommandExecutor.destroy();
+				} catch (Exception ex) {
+					log.warn("Cannot properly close cluster command executor", ex);
+				}
+			}
+		} else if (state.compareAndSet(State.STARTED, State.STOPPING)) {
 
-			try {
-				clusterCommandExecutor.destroy();
-			} catch (Exception ex) {
-				log.warn("Cannot properly close cluster command executor", ex);
+			resetConnection();
+
+			if (clusterCommandExecutor != null) {
+
+				try {
+					clusterCommandExecutor.destroy();
+				} catch (Exception ex) {
+					log.warn("Cannot properly close cluster command executor", ex);
+				}
 			}
-		}
 
-		dispose(connectionProvider);
-		dispose(reactiveConnectionProvider);
+			dispose(connectionProvider);
+			dispose(reactiveConnectionProvider);
 
-		try {
-			Duration quietPeriod = clientConfiguration.getShutdownQuietPeriod();
-			Duration timeout = clientConfiguration.getShutdownTimeout();
-			client.shutdown(quietPeriod.toMillis(), timeout.toMillis(), TimeUnit.MILLISECONDS);
-		} catch (Exception e) {
+			try {
```

Comment:
That code could go into `stop` and upon destroy it would be neat to reuse what `stop` does.